cpp/src/reshape/explode.cu

```diff
@@ -36,119 +36,225 @@
-    offsets + 1, [offsets] __device__(auto i) { return (i - offsets[0]) - 1; });
-  auto counting_iter = thrust::make_counting_iterator(0);
-
-  rmm::device_uvector<size_type> pos(include_pos ? sliced_child.size() : 0, stream, mr);
-
-  // This looks like an off-by-one bug, but what is going on here is that we need to reduce each
-  // result from `lower_bound` by 1 to build the correct gather map. This can be accomplished by
-  // skipping the first entry and using the result of `lower_bound` directly.
-  if (include_pos) {
-    thrust::transform(
-      rmm::exec_policy(stream),
-      counting_iter,
-      counting_iter + gather_map_indices.size(),
-      gather_map_indices.begin(),
-      [position_array = pos.data(), offsets_minus_one, offsets, offset_size = lc.size()] __device__(
-        auto idx) -> size_type {
-        auto lb_idx = thrust::lower_bound(
-                        thrust::seq, offsets_minus_one, offsets_minus_one + offset_size, idx) -
-                      offsets_minus_one;
-        position_array[idx] = idx - (offsets[lb_idx] - offsets[0]);
-        return lb_idx;
-      });
-  } else {
-    thrust::lower_bound(rmm::exec_policy(stream),
-                        offsets_minus_one,
-                        offsets_minus_one + lc.size(),
-                        counting_iter,
-                        counting_iter + gather_map_indices.size(),
-                        gather_map_indices.begin());
-  }
+using GatherType   = rmm::device_uvector<size_type>;
+using PositionType = rmm::device_uvector<size_type>;
 
+std::unique_ptr<table> build_table(table_view const& input_table,
+                                   size_type const explode_column_idx,
+                                   column_view const& sliced_child,
+                                   GatherType const& gather_map,
+                                   GatherType const& explode_col_gather_map,
+                                   PositionType position_array,
+                                   bool include_position,
+                                   rmm::cuda_stream_view stream,
+                                   rmm::mr::device_memory_resource* mr)
+{
   auto select_iter = thrust::make_transform_iterator(
     thrust::make_counting_iterator(0),
     [explode_column_idx](size_type i) { return i >= explode_column_idx ? i + 1 : i; });
   std::vector<size_type> selected_columns(select_iter, select_iter + input_table.num_columns() - 1);
 
-  auto gathered_table = cudf::detail::gather(input_table.select(selected_columns),
-                                             gather_map_indices.begin(),
-                                             gather_map_indices.end(),
-                                             cudf::out_of_bounds_policy::DONT_CHECK,
-                                             stream,
-                                             mr);
+  auto gathered_table = detail::gather(input_table.select(selected_columns),
```

Comment:
Seems like it'd be a nice improvement to `table_view::select` to add a variant that takes an iterator. The current version that takes a vector can then be made to call that.

Reply:
Implemented this in `table_view` and `table` to hopefully be useful to others. It does introduce another template in a well-used header though.